### ISSUE TITLE
ISTIO Sidecar performance impact Istio: 1.0.6. same observed in 1.2.3

### ISSUE THREAD

**Issue by nokiabaitanik** [istio/istio#15981]
**After side car Injected ,  performance dropped to half .  Even the request incoming is not at steady rate. over time it is degrading. istio version : 1.0.6**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[X ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior : After side car Injected ,  if performance  drops , it should not be more than 10%.  And performance should be steady , with time it should not degrade.***

**Following are steps our testing: **
  Without side car :
    1. client simulator (vegeta) to send GET/POST request -> application pod using pod service IP
     In statistics we can see it is able to reach 8k  tps for 8 hours run . it was stable for 8hours.
    2.  Side car attached to the pod.  (we have mTLS disabled)
      client simulator (vegeta) to send GET/POST request -> application pod using pod service IP.
     In statistics we can see it was reaching upto 4k . BUT problem is that it is degrading over time.
     it 8 hours end it was around 500.

**istio version : 1.0.6**

**Environment:**
   We are using kubernetes cluster.
   Deployment pod were given CPU: 4., Memory: 4Gi
   Application is using 2Gi Memory.
   So side car has 2 Gb memory to use.
 
**Environment where bug was observed openstack**

**Comment by howardjohn** [comment]
A couple things:

* I would recommend reading https://istio.io/blog/2019/performance-best-practices/

* Istio 1.0 is end of life now. We have made substantial performance improvements since then so I would encourage you to try out newer versions.

> Deployment pod were given CPU: 4., Memory: 4Gi
> Application is using 2Gi Memory.
> So side car has 2 Gb memory to use.

Kubernetes handles resources by container, not by pod. The sidecar has its own limits and requests assigned to it. I forgot what the defaults are on 1.0 since we have changed them since, but you can check with `kubectl get pod -oyaml`.

The performance getting worse over time suggests to me some component, likely mixer telemetry, is overloaded perhaps.



Also consider that if you have a low latency application, say 1ms, and you Envoy adds roughly 1ms of overhead, that means your throughput will be 50% of what it was without Envoy unless you increase the number of connections. More details are in the blog I posted above.

**Comment by nokiabaitanik** [comment]
For mixer we had " policyCheckFailOpen: true" . 
so even if mixer is over loaded traffic should get passed . as policyCheckFailOpen is set to true.
#policyCheckFailOpen allows traffic in cases when the mixer policy service cannot be reached.
#Default is false which means the traffic is denied when the client is unable to connect to Mixer.

Upgrading istio is option is there. Still we want to verify everything . then will go for upgrade. 

How to increase the cpu for only for side car . it is 10m . default value is taking ... 
I tried to update the deployment file of application. with the following :
```
annotations:
    sidecar.istio.io/proxyCPU: "1"
    sidecar.istio.io/proxyMemory: 1Gi
```
And restarted the pod , still it is showing default value.

**Comment by howardjohn** [comment]
> For mixer we had " policyCheckFailOpen: true" . 

This is only for mixer policy, mixer telemetry is also a factor. Even if it fails open, it can still impact latency though I believe, as it may take some time for the request to fail.

> How to increase the cpu for only for side car . it is 10m . default value is taking ...

For 1.0 I am not sure, we don't support that version anymore and I have never used it.

In 1.2 you can use https://istio.io/docs/reference/config/installation-options/#global-options 
 `global.proxy.resources.requests.cpu`


So are you using mixer policy at all?

**Comment by nokiabaitanik** [comment]
We have tried to disabled mixer .. but that also is giving same result. was there any memory leak related problem in 1.0.6 . any idea?

**Comment by howardjohn** [comment]
There have been many many performance problems addressed since 1.3. Without
more details about what problems you are seeing it's hard to say

**Comment by nokiabaitanik** [comment]
Okay. fine. we will update the istio version and check and if still it persists , will seek for help. 
I am closing the ticket now.
Thanks for the help and information.

**Comment by nokiabaitanik** [comment]
We updated the istio to 1.2.3 . but issue still persists . performance is degrading over time. 
without istio tps is going to 8k. with istio side car it is max going 2.5k .. 
But the main issue is it is degrading over time. even it is 2.5k ,it should be stable at 2.5k. 

For application container:
resources:
      limits:
        cpu: "2"
        memory: 3Gi
      requests:
        cpu: "2"
        memory: 3Gi
For istio side car :
resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: 100m
        memory: 128Mi

**Comment by howardjohn** [comment]
How did you install istio? Is mixer enabled? Is it using high CPU/hitting the CPU limit?

Did you read https://istio.io/blog/2019/performance-best-practices/?

Can you provide more info about cpu usage of all components, what you set up is?

**Comment by nokiabaitanik** [comment]
we used helm charts to install the istio . mixer was enabled . 
mixer:
  policy:
    # if policy is enabled the global.disablePolicyChecks has affect.
    enabled: true
  telemetry:
    enabled: true
Application was only 27% of cpu. 
"cpu usage of all components, what you set up is?" -- > what components you need . If you have commands , please share that.

![image](https://user-images.githubusercontent.com/47983805/62919694-19298500-bdc1-11e9-9106-70ab32148b2b.png)

**Comment by nokiabaitanik** [comment]
we have tried to disable mixer . but same thing.
disablePolicyCheck: true
mixerReportServer: disabled 

And apart for application and istio-proxy nothing is taking up the cpu.

**Comment by Asisranjan** [comment]
please refer https://github.com/istio/tools/blob/81cc22348059bb17ad9c2f571018e78780a1bbf5/perf/istio-install/values.yaml

**Comment by nokiabaitanik** [comment]
resource wise , I am not sure whether there is a problem or not . because cpu limit was not hitting earlier. 
and even for 500 tps it behaving same. But for 500 tps , it is stable for 2 hours . Then it starts dropping .
we are using the values.yaml what comes default with the istio helm charts. the cpu limis  is 2 and memory limit is 1Gi.
![image](https://user-images.githubusercontent.com/47983805/63011285-a2b28300-bea5-11e9-8dd9-352243765154.png)

**Comment by nokiabaitanik** [comment]
Please give an update. What others details you need?

**Comment by howardjohn** [comment]
Closed as dupe of #16370
